File: LeetCode/1372. Longest ZigZag Path in a Binary Tree/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def longestZigZag(self, root: TreeNode) -> int:
        def dfs(cur):
            if not cur:
                return 0, 0, 0
            ll, lr, lh = dfs(cur.left)
            rl, rr, rh = dfs(cur.right)
            return lr+1, rl+1, max(lr+1,rl+1,lh,rh)
        return dfs(root)[-1]-1

File: LeetCode/1372. Longest ZigZag Path in a Binary Tree/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
using tiii = tuple<int,int,int>;
class Solution {
public:
    tiii dfs(TreeNode *root){
        if (!root)
            return {0,0,0};
        auto [ll,lr,ld] = dfs(root->left);
        auto [rl,rr,rd] = dfs(root->right);
        int d = max(ld,rd);
        d = max(d,lr+1);
        d = max(d,rl+1);
        return {lr+1,rl+1,d};
    }
    int longestZigZag(TreeNode* root) {
        auto [l,r,d] = dfs(root);
        return d-1;
    }
};